### ISSUE TITLE
VM: Can we run code in parallel?

### ISSUE THREAD

**Issue by jochem-brouwer** [ethereumjs/ethereumjs-monorepo#959]
We have to figure out if there is a concurrency issue in case you try to run two VM calls in parallel. I'd assume this would mess up the state trie, as both calls point to the same `StateManager`, which also holds things like accessed storage slots for gas calculation, etc.

**Comment by alcuadrado** [comment]
This is an interesting topic. To me, it's totally unclear which Ethereumjs things can work in parallel. In Hardhat Network we have a global mutex just in case.

**Comment by jochem-brouwer** [comment]
We can't run VM in parallel, see the linked PR. 

How do we proceed here? Do we want to instantiate a lock to ensure that we don't run the VM in parallel? (For the time being)? As a follow-up we could add the feature to run the VM in parallel (so a single VM can run multiple things at the same time). I'd say to allow that we can create a `StateManagerPool` as it is likely that if we run each VM call on their own respective `StateManager` then the problem is solved. `StateManagerPool` returns a `StateManager` which is currently not being used from the pool, if none exists then create a new one.

**Comment by holgerd77** [comment]
Is this really a problem? Don't we want to execute sequentially on the block numbers anyhow and could therefore use a simple approach like (rough pseudo-code):

```typescript
let lastExecutedBlock = 0

if (blockToExecute.blockNumber == lastExecutedBlock) {
  await vm.runBlock(blockToExecute)
  lastExecutedBlock = blockToExecute.blockNumber
}
```

? 🤔

**Comment by jochem-brouwer** [comment]
Yes, in most settings this is not a problem, but if a consumer wants to use the VM in parallel and there are no guards in the VM which prevent the VM running in parallel and thus reporting wrong results, then the consumer will probably take some time to figure out that VM cannot be ran in parallel. I'd say that we should guard against this in the VM.

**Comment by alcuadrado** [comment]
I think adding a warning in the readme will also be ok. As a consumer, I'm ok with the vm not being being "thread safe", but not so much with it not being documented.

**Comment by jochem-brouwer** [comment]
I agree here, we should explicitly put this in documentation. I am also convinced that we don't need to add explicit guards here, if we communicate it to our consumers then it should be fine.

**Comment by holgerd77** [comment]
Have added this ("put this in documentation") to the tiny fixes issue #905 and will otherwise close here.
